File: backend/app/routers/leads.py
import logging
from typing import List, Optional

from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.deps import get_current_user, get_current_user_optional
from app.email_adapter import LeadNotification, get_email_adapter
from app.models import Lead, User
from app.schemas import LeadCreateRequest, LeadOut

logger = logging.getLogger("inspeckt.leads")

router = APIRouter(prefix="/api/v1/leads", tags=["leads"])

_VALID_KINDS = {"doesnt_fit", "license", "custom_bundle", "dispute", "escalation"}


@router.post("", response_model=LeadOut, status_code=status.HTTP_201_CREATED)
def create_lead(
    payload: LeadCreateRequest,
    db: Session = Depends(get_db),
    current_user: Optional[User] = Depends(get_current_user_optional),
):
    """Every AI-to-human hand-off in the product lands here: a product
    that doesn't fit any FSSAI category, a license/registration request, a
    dispute, a rule-level escalation, or a custom-bundle pricing enquiry.
    Works logged-out (the entry-point buttons on the homepage don't require
    an account first) as well as logged-in. See PLATFORM_PLAN_V2.md §3, §6.
    """
    kind = payload.kind if payload.kind in _VALID_KINDS else "doesnt_fit"
    record = Lead(
        user_id=current_user.id if current_user else None,
        kind=kind,
        contact_name=payload.contact_name or (current_user.full_name if current_user else None),
        contact_email=payload.contact_email,
        contact_phone=payload.contact_phone,
        company_name=payload.company_name or (current_user.company_name if current_user else None),
        message=payload.message,
        context_json=payload.context,
    )
    db.add(record)
    db.commit()
    db.refresh(record)

    # Every AI-to-human hand-off in the product depends on someone
    # actually finding out a lead exists — the DB row alone isn't enough
    # (see this router's own module comment). But the lead has already
    # been durably saved above: a founder who clicked "Talk to a
    # Consultant" must always get their 201 and their lead ID, even if
    # the notification email fails to send (misconfigured provider,
    # provider outage, etc.) — the alternative (failing the whole request
    # over a notification problem) would lose the lead's own creation for
    # no good reason. So this is deliberately best-effort: log and move
    # on, never raise.
    try:
        get_email_adapter().send_lead_notification(LeadNotification(
            kind=record.kind,
            contact_name=record.contact_name,
            contact_email=record.contact_email,
            contact_phone=record.contact_phone,
            company_name=record.company_name,
            message=record.message,
            lead_id=record.id,
        ))
    except Exception:
        logger.exception("Failed to send lead notification for lead %s (kind=%s)", record.id, record.kind)

    return LeadOut(id=record.id, kind=record.kind, status=record.status, created_at=record.created_at, message=record.message)


@router.get("", response_model=List[LeadOut])
def list_my_leads(db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    """Added for parity with the standalone frontend's existing
    listLeadsForUser() (frontend/js/leads.js) — a founder's dashboard has
    always shown "your consultant requests" (frontend/js/ui/dashboard.js's
    renderConsultRequests()); this endpoint didn't exist yet because
    nothing called it until the frontend was wired to this real backend.
    Auth-required (unlike POST above, which also works logged-out) since
    there's no way to scope "your own leads" without knowing who "you"
    are — a lead submitted while logged out (user_id is NULL) never shows
    up here, matching create_lead()'s own behavior of only stamping
    user_id when someone was actually logged in at submission time.
    """
    rows = (
        db.query(Lead)
        .filter(Lead.user_id == current_user.id)
        .order_by(Lead.created_at.desc())
        .all()
    )
    return [LeadOut(id=r.id, kind=r.kind, status=r.status, created_at=r.created_at, message=r.message) for r in rows]

File: frontend/js/leads.js
/**
 * Inspeckt — human hand-off capture, backed by the real FastAPI backend
 * (POST/GET /api/v1/leads — backend/app/routers/leads.py). Every "talk to
 * a consultant" moment in the product — a product that doesn't fit any
 * FSSAI category, a license/registration request, a custom-bundle
 * enquiry, a dispute, or a rule-level escalation from a report — is
 * submitted here. See PLATFORM_PLAN_V2.md §3 and §6.
 *
 * Used to write only to this browser's localStorage, with nothing behind
 * it at all — see git history for that version, and the long comment
 * that used to sit here about mailto: being "the one delivery mechanism
 * that works with no backend." The backend now actually emails the
 * consultant team on every submission (see backend/app/email_adapter.py's
 * ResendEmailAdapter, once EMAIL_PROVIDER=resend is configured), so the
 * mailto: fallback below is now a backup path — kept because it's a real,
 * zero-configuration way for a visitor to reach a human even if this
 * deployment's email sending isn't set up yet, not because it's the only
 * one.
 */
import { apiFetch } from "./api.js";

/**
 * @param {"doesnt_fit"|"license"|"custom_bundle"|"dispute"|"escalation"} kind
 * @param {{contactName?:string, contactEmail:string, contactPhone?:string, companyName?:string, message?:string, context?:object, userId?:string|null}} details
 * @returns {Promise<object>} the lead, including its server-assigned id —
 *   merges `details` back in (rather than only returning what the server
 *   echoes back) so buildConsultMailto() below has the contact fields it
 *   needs without a second round-trip; POST /api/v1/leads intentionally
 *   doesn't echo them back (see LeadOut in backend/app/schemas.py).
 */
export async function submitLead(kind, details) {
  const record = await apiFetch("/api/v1/leads", {
    method: "POST",
    // Works logged out too (get_current_user_optional server-side) — the
    // Authorization header is only attached by apiFetch when a session
    // token actually exists, matching this endpoint's own design.
    body: {
      kind,
      contact_name: details.contactName || null,
      contact_email: details.contactEmail,
      contact_phone: details.contactPhone || null,
      company_name: details.companyName || null,
      message: details.message || null,
      context: details.context || null,
    },
  });
  return { ...details, id: record.id, kind: record.kind, status: record.status, createdAt: record.created_at };
}

/** A logged-in founder's own consultant requests — see backend's
 * GET /api/v1/leads, which scopes this to the current account server-side
 * (a lead submitted logged out never has an owner to scope it to).
 * @returns {Promise<Array>}
 */
export async function listLeadsForUser(userId) {
  let rows;
  try {
    rows = await apiFetch("/api/v1/leads");
  } catch {
    return [];
  }
  return rows.map((r) => ({
    id: r.id, userId, kind: r.kind, status: r.status, message: r.message, createdAt: r.created_at,
  }));
}

// ---------------------------------------------------------------
// Real delivery, without relying on this deployment's email being set up
// ----------------------------------------------------------------
// A mailto: link is a delivery mechanism that works with zero backend
// configuration: it opens the *visitor's own* email client with the
// message pre-filled, and sending it is a real email from a real inbox to
// a real inbox — a useful fallback for a deployment that hasn't set
// EMAIL_PROVIDER=resend yet (see this file's header comment).
// CONSULT_INBOX_EMAIL must be a real, monitored address before this goes
// anywhere near real customers.
export const CONSULT_INBOX_EMAIL = "hello@inspecktfoodsolutions.com";

const LEAD_KIND_SUBJECT = {
  doesnt_fit: "New product — needs review",
  license: "License / registration request",
  custom_bundle: "Custom plan enquiry",
  dispute: "Dispute a report finding",
  escalation: "Report finding needs review",
  label_photo: "Label photo review request",
};

export function buildConsultMailto(lead) {
  const subject = `Inspeckt enquiry: ${LEAD_KIND_SUBJECT[lead.kind] || "Consultant request"} (ref ${lead.id})`;
  const body = [
    lead.message || "(no message entered)",
    "",
    `Name: ${lead.contactName || "—"}`,
    `Company: ${lead.companyName || "—"}`,
    `Phone: ${lead.contactPhone || "—"}`,
    `Email: ${lead.contactEmail || "—"}`,
    `Reference: ${lead.id}`,
  ].join("\n");
  return `mailto:${CONSULT_INBOX_EMAIL}?subject=${encodeURIComponent(subject)}&body=${encodeURIComponent(body)}`;
}
